catalog/tests/factories/github/__init__.py

```diff
@@ -152,3 +179,92 @@ def make_branch_protection(required_reviewers: int = 2) -> dict:
     ] = required_reviewers
 
     return branch_protection
+
+
+EventsConfig = namedtuple("EventsConfig", "event_name, days_since_previous_event")
+
+
+def make_events(origin_days_ago: int, events: list[str | EventsConfig]) -> list[dict]:
+    """
+    Create an issue events list with events starting at the origin.
+
+    The complexity of this fixture function is necessary to ease building
+    events list and testing the various states an events list can be in.
+
+    :param origin_days_ago: The number of days ago to treat as the origin
+    for the events list. If 0 then all events will be treated as being created the
+    same day and days lapsed configured for individual events will be ignored.
+    :param events: A list of mixed strings and (string, int) tuples. The string
+    must always be an event name. If a tuple, the int should be the number
+    of days to increment the event past the previous event.
+    """
+    result = []
+    previous_event_days_ago = origin_days_ago
+    for event_name in reversed(events):
+        days_since_previous_event = 0
+        if isinstance(event_name, tuple):
+            days_since_previous_event = event_name[1]
+            event_name = event_name[0]
+
+        days_ago = previous_event_days_ago + days_since_previous_event
```

Comment:
I think the actual bug with this function was with this line:

```suggestion
                # If the previous event was 2 days ago, and it has been one day since
                # the previous event, then this event was *1 day ago*
                days_ago = previous_event_days_ago - days_since_previous_event
```

Note that should be subtraction, not addition.

Reversing and unreversing the list does make the events happen in the right order chronologically, but completely messes up the actual dates (which is why it was causing failures for critical urgency). So for example here's an event list:

```
events_list = [
  "labeled",
  "review_requested",
  "convert_to_draft",
  ("ready_for_review", 1),
  "convert_to_draft",
  "ready_for_review"
]
```

The way this should read is: "The PR was opened (labeled/review requested) then converted to draft on the same day. One day later it was marked ready for review, converted to a draft, and marked ready for review again".

If I call `make_events` with this event list and `origin_days_ago` set to 2, it should make the list of events with the first event starting 2 days ago. If today is 2023-05-10, I expect results like this:

```
[{'name': 'labeled', 'created_at': '2023-05-08'},  # First event is 2 days ago
 {'name': 'review_requested', 'created_at': '2023-05-08'},
 {'name': 'convert_to_draft', 'created_at': '2023-05-08'},
 {'name': 'ready_for_review', 'created_at': '2023-05-09'}, # 1 day passed between the last event and this one
 {'name': 'convert_to_draft', 'created_at': '2023-05-09'},
 {'name': 'ready_for_review', 'created_at': '2023-05-09'}
]
```

If we don't do the list reversing and instead change that `days_ago` calculation, that is indeed what we get! 

With the list reversing instead, we end up with _the last issue being `origin_days_ago`, rather than the first issue_, and the days between events gets borked:
```
[{'name': 'labeled', 'created_at': '2023-05-05'}, # First event is now *five days ago*
 {'name': 'review_requested', 'created_at': '2023-05-05'},
 {'name': 'convert_to_draft', 'created_at': '2023-05-05'},
 {'name': 'ready_for_review', 'created_at': '2023-05-05'},
 {'name': 'convert_to_draft', 'created_at': '2023-05-08'}, # 3 days have passed before this event
 {'name': 'ready_for_review', 'created_at': '2023-05-08'}]
```

So this breaks the logic for testing the urgency because everything is too far in the past.

I think what we can undo all the changes to reverse/unreverse the list, change this calculation to subtraction, and then we should be able to undo all of the changes to the test parameters and it should just work!

Reply:
Ahhhh, I got it!! Amazing find!